File: apps/desktop/electron/research.ts
import * as fs from 'node:fs';

import type { ResearchAnswers, ResearchReport } from '@trackem/contracts';
export type { ResearchAnswers, ResearchReport } from '@trackem/contracts';
const EMPTY_ANSWERS: ResearchAnswers = { paidTools: 'unanswered', wouldPay: 'unanswered', useful: 'unanswered' };
export function normalizeAnswers(value: unknown): ResearchAnswers {
  const input = value && typeof value === 'object' ? value as Record<string, unknown> : {};
  const pick = (key: keyof ResearchAnswers, values: string[]) => typeof input[key] === 'string' && values.includes(input[key]) ? input[key] : 'unanswered';
  return {
    paidTools: pick('paidTools', ['one', 'two', 'three-plus']) as ResearchAnswers['paidTools'],
    wouldPay: pick('wouldPay', ['no', 'maybe', 'yes-3', 'yes-5', 'yes-10']) as ResearchAnswers['wouldPay'],
    useful: pick('useful', ['yes', 'no']) as ResearchAnswers['useful'],
  };
}

/** Local, opt-in study. Reports deliberately contain no identity, paths, quotas, or exact timestamps. */
export class LocalResearch {
  private startedAt = 0;
  private days = new Set<number>();
  private opens = 0;
  private lastOpen = 0;
  private answers = { ...EMPTY_ANSWERS };
  constructor(private file: string, private enabled: boolean) {
    if (!enabled) { this.clear(); return; }
    try {
      const data = JSON.parse(fs.readFileSync(file, 'utf8'));
      this.startedAt = typeof data.startedAt === 'number' && data.startedAt > 0 && data.startedAt <= Date.now() ? data.startedAt : Date.now();
      this.days = new Set(Array.isArray(data.days) ? data.days.filter((d: unknown) => Number.isInteger(d) && Number(d) >= 0 && Number(d) <= 89) : []);
      this.opens = Number.isSafeInteger(data.opens) && data.opens >= 0 ? data.opens : 0;
      this.answers = normalizeAnswers(data.answers);
    } catch { this.startedAt = Date.now(); }
  }
  setEnabled(enabled: boolean): void {
    if (enabled === this.enabled) return;
    this.enabled = enabled;
    this.clear();
  }
  recordOpen(now = Date.now()): void {
    if (!this.enabled || now - this.lastOpen < 60_000) return;
    if (!this.startedAt) this.startedAt = now;
    const day = Math.floor((now - this.startedAt) / 86_400_000);
    if (day > 89) return;
    this.lastOpen = now;
    this.days.add(day);
    this.opens++;
    this.persist();
  }
  save(value: unknown): ResearchReport {
    if (this.enabled) { this.answers = normalizeAnswers(value); this.persist(); }
    return this.report();
  }
  clear(): ResearchReport {
    this.startedAt = this.enabled ? Date.now() : 0;
    this.days.clear(); this.opens = 0; this.lastOpen = 0; this.answers = { ...EMPTY_ANSWERS };
    fs.rmSync(this.file, { force: true });
    return this.report();
  }
  report(): ResearchReport {
    return { version: 1, enabled: this.enabled, daysSinceStart: this.startedAt ? Math.max(0, Math.floor((Date.now() - this.startedAt) / 86_400_000)) : 0, activeDays: [...this.days].sort((a, b) => a - b), opens: this.opens, answers: { ...this.answers } };
  }
  private persist(): void {
    fs.writeFileSync(this.file, JSON.stringify({ startedAt: this.startedAt, days: [...this.days], opens: this.opens, answers: this.answers }), { mode: 0o600 });
    fs.chmodSync(this.file, 0o600);
  }
}

File: src-tauri/src/research.rs
use crate::types::{ResearchAnswers, ResearchReport};
use serde_json::{json, Value};
use std::collections::BTreeSet;
use std::fs;
use std::path::{Path, PathBuf};

#[derive(Debug)]
pub struct LocalResearch {
    file: PathBuf,
    enabled: bool,
    started_at: i64,
    days: BTreeSet<u64>,
    opens: u64,
    last_open: i64,
    answers: ResearchAnswers,
}

impl LocalResearch {
    pub fn load(file: PathBuf, enabled: bool) -> Self {
        let now = crate::forecast::now_millis();
        let mut research = Self { file, enabled, started_at: 0, days: BTreeSet::new(), opens: 0, last_open: 0, answers: empty_answers() };
        if !enabled {
            research.clear();
            return research;
        }
        if let Ok(contents) = fs::read_to_string(&research.file) {
            if let Ok(data) = serde_json::from_str::<Value>(&contents) {
                research.started_at = data.get("startedAt").and_then(Value::as_i64).filter(|at| *at > 0 && *at <= now).unwrap_or(now);
                research.days = data.get("days").and_then(Value::as_array).into_iter().flatten().filter_map(Value::as_u64).filter(|day| *day <= 89).collect();
                research.opens = data.get("opens").and_then(Value::as_u64).unwrap_or(0);
                research.answers = normalize_answers(data.get("answers").unwrap_or(&Value::Null));
            }
        }
        if research.started_at == 0 {
            research.started_at = now;
        }
        research
    }

    pub fn set_enabled(&mut self, enabled: bool) {
        if enabled == self.enabled { return }
        self.enabled = enabled;
        self.clear();
    }

    pub fn record_open(&mut self, now: i64) {
        if !self.enabled || now - self.last_open < 60_000 { return }
        if self.started_at == 0 { self.started_at = now }
        let day = ((now - self.started_at) / 86_400_000).max(0) as u64;
        if day > 89 { return }
        self.last_open = now;
        self.days.insert(day);
        self.opens += 1;
        self.persist();
    }

    pub fn save(&mut self, value: &Value) -> ResearchReport {
        if self.enabled {
            self.answers = normalize_answers(value);
            self.persist();
        }
        self.report()
    }

    pub fn clear(&mut self) -> ResearchReport {
        self.started_at = if self.enabled { crate::forecast::now_millis() } else { 0 };
        self.days.clear();
        self.opens = 0;
        self.last_open = 0;
        self.answers = empty_answers();
        let _ = fs::remove_file(&self.file);
        self.report()
    }

    pub fn report(&self) -> ResearchReport {
        ResearchReport {
            version: 1,
            enabled: self.enabled,
            days_since_start: if self.started_at == 0 { 0 } else { ((crate::forecast::now_millis() - self.started_at).max(0) / 86_400_000) as u64 },
            active_days: self.days.iter().copied().collect(),
            opens: self.opens,
            answers: self.answers.clone(),
        }
    }

    fn persist(&self) {
        let Some(parent) = self.file.parent() else { return };
        if fs::create_dir_all(parent).is_err() { return }
        let value = json!({ "startedAt": self.started_at, "days": self.days, "opens": self.opens, "answers": self.answers });
        if let Ok(contents) = serde_json::to_vec(&value) {
            if fs::write(&self.file, contents).is_ok() {
                set_private_permissions(&self.file);
            }
        }
    }
}

fn empty_answers() -> ResearchAnswers {
    ResearchAnswers { paid_tools: "unanswered".to_string(), would_pay: "unanswered".to_string(), useful: "unanswered".to_string() }
}

pub fn normalize_answers(value: &Value) -> ResearchAnswers {
    let paid = value.get("paidTools").and_then(Value::as_str).unwrap_or("");
    let would = value.get("wouldPay").and_then(Value::as_str).unwrap_or("");
    let useful = value.get("useful").and_then(Value::as_str).unwrap_or("");
    ResearchAnswers {
        paid_tools: if ["one", "two", "three-plus"].contains(&paid) { paid.to_string() } else { "unanswered".to_string() },
        would_pay: if ["no", "maybe", "yes-3", "yes-5", "yes-10"].contains(&would) { would.to_string() } else { "unanswered".to_string() },
        useful: if ["yes", "no"].contains(&useful) { useful.to_string() } else { "unanswered".to_string() },
    }
}

#[cfg(unix)]
fn set_private_permissions(file: &Path) {
    use std::os::unix::fs::PermissionsExt;
    let _ = fs::set_permissions(file, fs::Permissions::from_mode(0o600));
}

#[cfg(not(unix))]
fn set_private_permissions(_file: &Path) {}
